keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -2619,6 +2619,14 @@ Beta:
   - Automatically manages versions of mandatory/suggested configuration
   - Discrimates paginated LIST requests
 
+GA:
+
+- Satisfaction with LIST and WATCH support
+- APF allows us to disable client-side rate limiting (or we know the reason why not)
```

Comment:
If we cannot disable client-side rate limiting, why would we consider the feature complete?

Reply:
In my opinion there are two aspects:
- whether without client-side rate-limiting the kube-apiserver (and etcd) are falling over
- whether without client-side rate-limiting some other aspects of the systems are falling over because they are not able to keep up with load and accumulating huge backlog

Let's make a specific example. Let's say that endpoints controller can keep up with 100pods/s throughput in large enough clusters. Now, if we remove client-side rate-limiting completely, then e.g. scheduler would be able to scheduler say 500pods/s and endpoints controller will be accumulating backlog - thus network programming latency will go extremely high.

From that perspective, I think that APF does that job (and we would be able to get rid of client-side rate-limiting - there is nothing more to do there), but I still will be reluctant to actually getting rid of it from all components to avoid hurting other aspects of the system. But the improvements needed would actually be in completely different components, so shouldn't block APF graduation.

Reply:
That's a compelling reason. I'd like for that explanation or something similar to be expressed here.  Priority and fairness, with client-side rate limiting disabled, must be sufficient for the kube-apiserver to survive. If other components require additional client-side rate limiting, that will not stop our GA, but the kube-apiserver must survive without it.

Reply:
@MikeSpreitzer - can you please extend the text to incorporate it?

Reply:
Interesting distinction here.  In short, this harkens back to the point I have been making in other contexts: Kubernetes can be seen as two layers.  The lower layer is an extensible API service, and the higher layer is built on that API service and consists of the resource definitions and controllers for managing containerized workload.  APF is focused on the lower layer, but Kubernetes will not be safe from overloads until the higher layer also has protections.

Reply:
I agree with the theory but I'm skeptical that we actually have some slow controller like this?

I'm also skeptical that this is how people would choose to solve a problem -- I've heard of tons of latency problems and I can't recall ever hearing someone suggest slowing down everything else in the system so that the slow thing can keep up.

And if you really want to slow down the e.g. scheduler to keep endpoints from looking bad, APF does let you do that...

Reply:
Longer-term I agree with that.

Shorter-term I actually disagree.
Slow networking controllers (that complete can't keep up with other controllers) can actually cause an outage to your applications.
e.g. imagine that we can do a rolling upgrade super fast, but networking doesn't keep up and when the last old pod is actually deleted, the newer ones are not yet added to the LB mechanism. That means a completely outage of your service.

So while I'm definitely all for disabling client-side rate-limits eventually, I'm lacking a lot of confidence here and I would be opposed doing that before we prove it works. And at the same time, blocking P&F graduation on that doesn't sound desired.

Reply:
Wasn't there a pod readiness thing done a while ago to address that load balancer scenario?

Reply:
That doesn't necessary mean everyone is using that. Recommended patterns are not necessarily quickly adopted by many users.